File: doctor/invariants/presets.py
"""Preset static invariants for IDE agent efficiency attribution."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable

Severity = str  # low | medium | high | critical
Category = str  # context | tool | mcp | skill


@dataclass
class Violation:
    invariant_id: str
    category: Category
    step_index: int
    severity: Severity
    message: str
    evidence: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "invariant_id": self.invariant_id,
            "category": self.category,
            "step_index": self.step_index,
            "severity": self.severity,
            "message": self.message,
            "evidence": self.evidence,
        }


@dataclass
class Invariant:
    invariant_id: str
    category: Category
    description: str
    check_fn: Callable[[dict[str, Any], list[dict[str, Any]]], list[Violation]]


def _inv_ctx_001(traj: dict[str, Any], steps: list[dict[str, Any]]) -> list[Violation]:
    """Single step with too many tool calls (>8)."""
    out = []
    for s in steps:
        tc = s.get("telemetry", {}).get("tool_count", 0)
        if tc > 8:
            out.append(Violation(
                "INV-CTX-001", "context", s["index"], "medium",
                f"Step {s['index']} has {tc} tool calls in one turn (threshold 8)",
                {"tool_count": tc},
            ))
    return out


def _inv_ctx_002(traj: dict[str, Any], steps: list[dict[str, Any]]) -> list[Violation]:
    """Excessive Read operations across session (>40)."""
    total_reads = sum(s.get("telemetry", {}).get("read_count", 0) for s in steps)
    if total_reads > 40:
        return [Violation(
            "INV-CTX-002", "context", steps[-1]["index"] if steps else 0, "high",
            f"Session has {total_reads} Read operations (threshold 40)",
            {"total_reads": total_reads},
        )]
    return []


def _inv_ctx_003(traj: dict[str, Any], steps: list[dict[str, Any]]) -> list[Violation]:
    """Agent self-spin: many assistant steps per user turn."""
    if not steps:
        return []
    user_turns = traj.get("metadata", {}).get("user_turns", 1) or 1
    ratio = len(steps) / max(user_turns, 1)
    if ratio > 12:
        return [Violation(
            "INV-CTX-003", "context", steps[len(steps) // 2]["index"], "high",
            f"High assistant/user ratio {ratio:.1f}:1 ({len(steps)} steps / {user_turns} user turns)",
            {"step_count": len(steps), "user_turns": user_turns, "ratio": round(ratio, 2)},
        )]
    return []


def _inv_tool_001(traj: dict[str, Any], steps: list[dict[str, Any]]) -> list[Violation]:
    """Repeated Grep pattern >= 3 times."""
    from collections import Counter
    patterns: Counter[str] = Counter()
    first_step: dict[str, int] = {}
    for s in steps:
        for pat in s.get("telemetry", {}).get("grep_patterns", []):
            patterns[pat] += 1
            first_step.setdefault(pat, s["index"])
    out = []
    for pat, cnt in patterns.items():
        if cnt >= 3:
            out.append(Violation(
                "INV-TOOL-001", "tool", first_step[pat], "medium",
                f"Grep pattern repeated {cnt} times",
                {"pattern": pat[:200], "count": cnt},
            ))
    return out


def _inv_tool_002(traj: dict[str, Any], steps: list[dict[str, Any]]) -> list[Violation]:
    """Repeated Shell command >= 2 times."""
    from collections import Counter
    cmds: Counter[str] = Counter()
    first_step: dict[str, int] = {}
    for s in steps:
        for cmd in s.get("telemetry", {}).get("shell_cmds", []):
            norm = " ".join(cmd.split())[:500]
            cmds[norm] += 1
            first_step.setdefault(norm, s["index"])
    out = []
    for cmd, cnt in cmds.items():
        if cnt >= 2:
            out.append(Violation(
                "INV-TOOL-002", "tool", first_step[cmd], "medium",
                f"Shell command repeated {cnt} times",
                {"command": cmd[:300], "count": cnt},
            ))
    return out


def _inv_tool_003(traj: dict[str, Any], steps: list[dict[str, Any]]) -> list[Violation]:
    """Many harness test retries (harness test run >= 4)."""
    harness_runs = 0
    first_idx = 0
    for s in steps:
        for cmd in s.get("telemetry", {}).get("shell_cmds", []):
            if "harness test run" in cmd:
                harness_runs += 1
                if first_idx == 0:
                    first_idx = s["index"]
    if harness_runs >= 4:
        return [Violation(
            "INV-TOOL-003", "tool", first_idx or 1, "high",
            f"Harness test run invoked {harness_runs} times (trial-and-error loop)",
            {"harness_test_runs": harness_runs},
        )]
    return []


def _inv_mcp_001(traj: dict[str, Any], steps: list[dict[str, Any]]) -> list[Violation]:
    """MCP-heavy session: MCP calls > 30% of tools and total tools > 50."""
    total_tools = sum(s.get("telemetry", {}).get("tool_count", 0) for s in steps)
    mcp_calls = sum(s.get("telemetry", {}).get("mcp_count", 0) for s in steps)
    if total_tools > 50 and mcp_calls / max(total_tools, 1) > 0.3:
        return [Violation(
            "INV-MCP-001", "mcp", steps[-1]["index"] if steps else 0, "medium",
            f"MCP calls {mcp_calls}/{total_tools} ({100*mcp_calls/max(total_tools,1):.0f}%) — MCP-heavy exploration",
            {"mcp_calls": mcp_calls, "total_tools": total_tools},
        )]
    return []


def _inv_mcp_002(traj: dict[str, Any], steps: list[dict[str, Any]]) -> list[Violation]:
    """Excessive MCP tool usage (>100) suggesting log/query thrashing."""
    mcp_calls = sum(s.get("telemetry", {}).get("mcp_count", 0) for s in steps)
    if mcp_calls > 100:
        return [Violation(
            "INV-MCP-002", "mcp", steps[len(steps)//2]["index"], "high",
            f"Excessive MCP invocations: {mcp_calls} (possible log/query thrashing)",
            {"mcp_calls": mcp_calls},
        )]
    return []


def _inv_skill_001(traj: dict[str, Any], steps: list[dict[str, Any]]) -> list[Violation]:
    """Skill read but continued heavy exploration without deliverable writes."""
    skill_read = any(s.get("telemetry", {}).get("skill_reads") for s in steps)
    total_read = sum(s.get("telemetry", {}).get("read_count", 0) for s in steps)
    total_grep = sum(s.get("telemetry", {}).get("grep_count", 0) for s in steps)
    write_count = sum(
        1 for s in steps
        for tn in s.get("telemetry", {}).get("tool_names", [])
        if tn in ("Write", "StrReplace")
    )
    if skill_read and (total_read + total_grep) > 30 and write_count < 3:
        return [Violation(
            "INV-SKILL-001", "skill", steps[-1]["index"], "medium",
            f"Skill loaded but exploration-heavy ({total_read} reads, {total_grep} greps, {write_count} writes)",
            {"reads": total_read, "greps": total_grep, "writes": write_count},
        )]
    return []


def _inv_skill_002(traj: dict[str, Any], steps: list[dict[str, Any]]) -> list[Violation]:
    """Harness task without reading harness skill."""
    instr = (traj.get("instruction") or "").lower()
    if "harness" not in instr:
        return []
    harness_skill = any(
        "harness" in p.lower() and "skill" in p.lower()
        for s in steps
        for p in s.get("telemetry", {}).get("skill_reads", [])
    )
    harness_skill_read = any(
        "/harness/SKILL.md" in p or "skills/harness" in p.lower()
        for s in steps
        for p in s.get("telemetry", {}).get("skill_reads", [])
    )
    if not harness_skill_read and not harness_skill:
        # check if Read path contains harness skill
        read_harness = any(
            "harness/SKILL.md" in p
            for s in steps
            for p in s.get("telemetry", {}).get("read_paths", [])
        )
        if not read_harness:
            return [Violation(
                "INV-SKILL-002", "skill", 1, "low",
                "Task mentions harness but harness SKILL.md was not read early",
                {"instruction_snippet": instr[:200]},
            )]
    return []


def _inv_ctx_004(traj: dict[str, Any], steps: list[dict[str, Any]]) -> list[Violation]:
    """Scope creep: many user turns (>=20) with low write ratio."""
    user_turns = traj.get("metadata", {}).get("user_turns", 0)
    writes = sum(
        1 for s in steps
        for tn in s.get("telemetry", {}).get("tool_names", [])
        if tn in ("Write", "StrReplace")
    )
    if user_turns >= 20 and writes < 15:
        return [Violation(
            "INV-CTX-004", "context", steps[-1]["index"] if steps else 0, "high",
            f"Scope creep: {user_turns} user turns but only {writes} write/edit ops",
            {"user_turns": user_turns, "writes": writes},
        )]
    return []


PRESET_INVARIANTS: list[Invariant] = [
    Invariant("INV-CTX-001", "context", "Too many tools per step", _inv_ctx_001),
    Invariant("INV-CTX-002", "context", "Excessive Read ops", _inv_ctx_002),
    Invariant("INV-CTX-003", "context", "High assistant/user ratio", _inv_ctx_003),
    Invariant("INV-CTX-004", "context", "Scope creep low delivery", _inv_ctx_004),
    Invariant("INV-TOOL-001", "tool", "Repeated Grep", _inv_tool_001),
    Invariant("INV-TOOL-002", "tool", "Repeated Shell", _inv_tool_002),
    Invariant("INV-TOOL-003", "tool", "Harness retry loop", _inv_tool_003),
    Invariant("INV-MCP-001", "mcp", "MCP-heavy session", _inv_mcp_001),
    Invariant("INV-MCP-002", "mcp", "MCP thrashing", _inv_mcp_002),
    Invariant("INV-SKILL-001", "skill", "Skill read but over-explore", _inv_skill_001),
    Invariant("INV-SKILL-002", "skill", "Missing harness skill", _inv_skill_002),
]


def export_static_invariants() -> list[dict[str, Any]]:
    return [
        {"invariant_id": inv.invariant_id, "category": inv.category, "description": inv.description}
        for inv in PRESET_INVARIANTS
    ]

File: src/invariants/presets.ts
import type { Category, Severity, TrajectoryIR, TrajectoryStep, Violation } from "../types/index.js";

type CheckFn = (traj: TrajectoryIR, steps: TrajectoryStep[]) => Violation[];

interface Invariant {
  invariant_id: string;
  category: Category;
  description: string;
  check: CheckFn;
}

function v(
  id: string,
  category: Category,
  step: number,
  severity: Severity,
  message: string,
  evidence: Record<string, unknown> = {}
): Violation {
  return { invariant_id: id, category, step_index: step, severity, message, evidence };
}

const invCtx001: CheckFn = (_t, steps) => {
  const out: Violation[] = [];
  for (const s of steps) {
    const tc = s.telemetry.tool_count;
    if (tc > 8) {
      out.push(v("INV-CTX-001", "context", s.index, "medium", `Step ${s.index} has ${tc} tool calls in one turn (threshold 8)`, { tool_count: tc }));
    }
  }
  return out;
};

const invCtx002: CheckFn = (_t, steps) => {
  const total = steps.reduce((n, s) => n + s.telemetry.read_count, 0);
  if (total > 40) {
    return [v("INV-CTX-002", "context", steps.at(-1)?.index ?? 0, "high", `Session has ${total} Read operations (threshold 40)`, { total_reads: total })];
  }
  return [];
};

const invCtx003: CheckFn = (traj, steps) => {
  if (!steps.length) return [];
  const userTurns = traj.metadata.user_turns ?? 1;
  const ratio = steps.length / Math.max(userTurns, 1);
  if (ratio > 12) {
    return [v("INV-CTX-003", "context", steps[Math.floor(steps.length / 2)]!.index, "high",
      `High assistant/user ratio ${ratio.toFixed(1)}:1 (${steps.length} steps / ${userTurns} user turns)`,
      { step_count: steps.length, user_turns: userTurns, ratio: Math.round(ratio * 100) / 100 })];
  }
  return [];
};

const invCtx004: CheckFn = (traj, steps) => {
  const userTurns = traj.metadata.user_turns ?? 0;
  let writes = 0;
  for (const s of steps) {
    for (const tn of s.telemetry.tool_names) {
      if (tn === "Write" || tn === "StrReplace") writes++;
    }
  }
  if (userTurns >= 20 && writes < 15) {
    return [v("INV-CTX-004", "context", steps.at(-1)?.index ?? 0, "high",
      `Scope creep: ${userTurns} user turns but only ${writes} write/edit ops`, { user_turns: userTurns, writes })];
  }
  return [];
};

const invTool001: CheckFn = (_t, steps) => {
  const patterns = new Map<string, number>();
  const firstStep = new Map<string, number>();
  for (const s of steps) {
    for (const pat of s.telemetry.grep_patterns) {
      patterns.set(pat, (patterns.get(pat) ?? 0) + 1);
      if (!firstStep.has(pat)) firstStep.set(pat, s.index);
    }
  }
  const out: Violation[] = [];
  for (const [pat, cnt] of patterns) {
    if (cnt >= 3) {
      out.push(v("INV-TOOL-001", "tool", firstStep.get(pat)!, "medium", `Grep pattern repeated ${cnt} times`, { pattern: pat.slice(0, 200), count: cnt }));
    }
  }
  return out;
};

const invTool002: CheckFn = (_t, steps) => {
  const cmds = new Map<string, number>();
  const firstStep = new Map<string, number>();
  for (const s of steps) {
    for (const cmd of s.telemetry.shell_cmds) {
      const norm = cmd.split(/\s+/).join(" ").slice(0, 500);
      cmds.set(norm, (cmds.get(norm) ?? 0) + 1);
      if (!firstStep.has(norm)) firstStep.set(norm, s.index);
    }
  }
  const out: Violation[] = [];
  for (const [cmd, cnt] of cmds) {
    if (cnt >= 2) {
      out.push(v("INV-TOOL-002", "tool", firstStep.get(cmd)!, "medium", `Shell command repeated ${cnt} times`, { command: cmd.slice(0, 300), count: cnt }));
    }
  }
  return out;
};

const invTool003: CheckFn = (_t, steps) => {
  let harnessRuns = 0;
  let firstIdx = 0;
  for (const s of steps) {
    for (const cmd of s.telemetry.shell_cmds) {
      if (cmd.includes("harness test run")) {
        harnessRuns++;
        if (!firstIdx) firstIdx = s.index;
      }
    }
  }
  if (harnessRuns >= 4) {
    return [v("INV-TOOL-003", "tool", firstIdx || 1, "high", `Harness test run invoked ${harnessRuns} times (trial-and-error loop)`, { harness_test_runs: harnessRuns })];
  }
  return [];
};

const invTool004: CheckFn = (_t, steps) => {
  const out: Violation[] = [];
  for (const s of steps) {
    for (const sub of s.substeps) {
      const dur = sub.execution?.duration_ms;
      if (dur != null && dur >= 120_000) {
        out.push(v("INV-TOOL-004", "tool", s.index, "high",
          `Slow tool ${sub.tool_name}: ${Math.round(dur / 1000)}s at substep ${sub.sub_index}`,
          { tool: sub.tool_name, duration_ms: dur, sub_index: sub.sub_index }));
      }
    }
  }
  return out;
};

const invTool005: CheckFn = (_t, steps) => {
  const out: Violation[] = [];
  for (const s of steps) {
    for (const sub of s.substeps) {
      const tokens = sub.execution?.output_tokens ?? 0;
      if (tokens >= 50_000) {
        out.push(v("INV-TOOL-005", "tool", s.index, "high",
          `Bloated tool output ${sub.tool_name}: ~${tokens.toLocaleString()} tokens`,
          { tool: sub.tool_name, output_tokens: tokens, sub_index: sub.sub_index }));
      }
    }
  }
  return out;
};

const invTool006: CheckFn = (_t, steps) => {
  const totalMs = steps.reduce((n, s) => n + (s.telemetry.tool_duration_ms ?? 0), 0);
  if (totalMs >= 30 * 60_000) {
    return [v("INV-TOOL-006", "tool", steps.at(-1)?.index ?? 0, "high",
      `Session tool wall time ~${Math.round(totalMs / 60000)} min (threshold 30 min)`,
      { total_tool_duration_ms: totalMs })];
  }
  return [];
};

const invTool007: CheckFn = (_t, steps) => {
  let readTokens = 0;
  let readCount = 0;
  for (const s of steps) {
    for (const sub of s.substeps) {
      if (sub.tool_name === "Read") {
        readTokens += sub.execution?.output_tokens ?? 0;
        readCount++;
      }
    }
  }
  const avg = readCount ? readTokens / readCount : 0;
  if (readCount >= 20 && avg >= 5000) {
    return [v("INV-TOOL-007", "tool", steps[Math.floor(steps.length / 2)]!.index, "medium",
      `Read output bloat: avg ~${Math.round(avg).toLocaleString()} tokens across ${readCount} reads`,
      { avg_read_tokens: Math.round(avg), read_count: readCount })];
  }
  return [];
};

const invMcp001: CheckFn = (_t, steps) => {
  const totalTools = steps.reduce((n, s) => n + s.telemetry.tool_count, 0);
  const mcpCalls = steps.reduce((n, s) => n + s.telemetry.mcp_count, 0);
  if (totalTools > 50 && mcpCalls / Math.max(totalTools, 1) > 0.3) {
    return [v("INV-MCP-001", "mcp", steps.at(-1)?.index ?? 0, "medium",
      `MCP calls ${mcpCalls}/${totalTools} (${Math.round(100 * mcpCalls / totalTools)}%) — MCP-heavy exploration`,
      { mcp_calls: mcpCalls, total_tools: totalTools })];
  }
  return [];
};

const invMcp002: CheckFn = (_t, steps) => {
  const mcpCalls = steps.reduce((n, s) => n + s.telemetry.mcp_count, 0);
  if (mcpCalls > 100) {
    return [v("INV-MCP-002", "mcp", steps[Math.floor(steps.length / 2)]!.index, "high",
      `Excessive MCP invocations: ${mcpCalls} (possible log/query thrashing)`, { mcp_calls: mcpCalls })];
  }
  return [];
};

const invSkill001: CheckFn = (_t, steps) => {
  const skillRead = steps.some((s) => s.telemetry.skill_reads.length > 0);
  const totalRead = steps.reduce((n, s) => n + s.telemetry.read_count, 0);
  const totalGrep = steps.reduce((n, s) => n + s.telemetry.grep_count, 0);
  let writeCount = 0;
  for (const s of steps) {
    for (const tn of s.telemetry.tool_names) {
      if (tn === "Write" || tn === "StrReplace") writeCount++;
    }
  }
  if (skillRead && totalRead + totalGrep > 30 && writeCount < 3) {
    return [v("INV-SKILL-001", "skill", steps.at(-1)?.index ?? 0, "medium",
      `Skill loaded but exploration-heavy (${totalRead} reads, ${totalGrep} greps, ${writeCount} writes)`,
      { reads: totalRead, greps: totalGrep, writes: writeCount })];
  }
  return [];
};

const invSkill002: CheckFn = (traj, steps) => {
  const instr = (traj.instruction ?? "").toLowerCase();
  if (!instr.includes("harness")) return [];
  const readHarness = steps.some((s) =>
    s.telemetry.read_paths.some((p) => p.includes("harness/SKILL.md")) ||
    s.telemetry.skill_reads.some((p) => p.includes("harness/SKILL.md") || p.toLowerCase().includes("skills/harness")) ||
    s.telemetry.shell_cmds.some((c) => c.includes("harness/SKILL.md") || c.includes("skills/harness"))
  );
  if (!readHarness) {
    return [v("INV-SKILL-002", "skill", 1, "low", "Task mentions harness but harness SKILL.md was not read early", { instruction_snippet: instr.slice(0, 200) })];
  }
  return [];
};

const invCodex001: CheckFn = (traj, steps) => {
  const eff = traj.metadata.tool_efficiency as { background_sessions?: Array<{ session_id: number; poll_count: number; total_wall_ms: number; aggregated_ms?: number; command: string; first_step?: number }> } | undefined;
  const sessions = eff?.background_sessions ?? [];
  const out: Violation[] = [];
  for (const bg of sessions) {
    const wall = bg.aggregated_ms ?? bg.total_wall_ms;
    if (bg.poll_count >= 3 && wall >= 60_000) {
      out.push(v("INV-CODEX-001", "tool", bg.first_step ?? steps[0]?.index ?? 1, "high",
        `Background exec session ${bg.session_id} polled ${bg.poll_count} times (~${Math.round(wall / 1000)}s wall)`,
        { session_id: bg.session_id, poll_count: bg.poll_count, total_wall_ms: wall, command: bg.command.slice(0, 200) }));
    }
  }
  return out;
};

const invCodex002: CheckFn = (traj, steps) => {
  const eff = traj.metadata.tool_efficiency as { thinking_gaps_ms?: Array<{ after_step: number; gap_ms: number; label: string }> } | undefined;
  const gaps = eff?.thinking_gaps_ms ?? [];
  const out: Violation[] = [];
  for (const gap of gaps) {
    if (gap.gap_ms >= 120_000) {
      out.push(v("INV-CODEX-002", "context", gap.after_step, "high",
        `Long idle gap ${Math.round(gap.gap_ms / 1000)}s before next step (model/tool wait)`,
        { gap_ms: gap.gap_ms, label: gap.label.slice(0, 200) }));
    }
  }
  return out;
};

const invCodex003: CheckFn = (_t, steps) => {
  let discovery = 0;
  let envUp = 0;
  let firstDiscovery = 0;
  for (const s of steps.slice(0, 15)) {
    for (const cmd of s.telemetry.shell_cmds) {
      const lower = cmd.toLowerCase();
      if (/^(find|rg|grep|ls|sed|cat)\s/.test(lower) || lower.includes(" --help") || lower.includes("worktree list")) {
        discovery++;
        if (!firstDiscovery) firstDiscovery = s.index;
      }
      if (lower.includes("harness env up") || lower.includes("pnpm install") || lower.includes("pnpm run serve")) {
        envUp++;
      }
    }
  }
  if (discovery >= 8 && envUp === 0) {
    return [v("INV-CODEX-003", "context", firstDiscovery || 1, "medium",
      `Extended discovery (${discovery} search/help cmds) before environment bootstrap`,
      { discovery_cmds: discovery, env_up_cmds: envUp })];
  }
  return [];
};

export const PRESET_INVARIANTS: Invariant[] = [
  { invariant_id: "INV-CTX-001", category: "context", description: "Too many tools per step", check: invCtx001 },
  { invariant_id: "INV-CTX-002", category: "context", description: "Excessive Read ops", check: invCtx002 },
  { invariant_id: "INV-CTX-003", category: "context", description: "High assistant/user ratio", check: invCtx003 },
  { invariant_id: "INV-CTX-004", category: "context", description: "Scope creep low delivery", check: invCtx004 },
  { invariant_id: "INV-TOOL-001", category: "tool", description: "Repeated Grep", check: invTool001 },
  { invariant_id: "INV-TOOL-002", category: "tool", description: "Repeated Shell", check: invTool002 },
  { invariant_id: "INV-TOOL-003", category: "tool", description: "Harness retry loop", check: invTool003 },
  { invariant_id: "INV-TOOL-004", category: "tool", description: "Slow single tool execution", check: invTool004 },
  { invariant_id: "INV-TOOL-005", category: "tool", description: "Bloated tool output", check: invTool005 },
  { invariant_id: "INV-TOOL-006", category: "tool", description: "Excessive total tool wall time", check: invTool006 },
  { invariant_id: "INV-TOOL-007", category: "tool", description: "Read output bloat", check: invTool007 },
  { invariant_id: "INV-MCP-001", category: "mcp", description: "MCP-heavy session", check: invMcp001 },
  { invariant_id: "INV-MCP-002", category: "mcp", description: "MCP thrashing", check: invMcp002 },
  { invariant_id: "INV-SKILL-001", category: "skill", description: "Skill read but over-explore", check: invSkill001 },
  { invariant_id: "INV-SKILL-002", category: "skill", description: "Missing harness skill", check: invSkill002 },
  { invariant_id: "INV-CODEX-001", category: "tool", description: "Codex background exec polling", check: invCodex001 },
  { invariant_id: "INV-CODEX-002", category: "context", description: "Codex long thinking gap", check: invCodex002 },
  { invariant_id: "INV-CODEX-003", category: "context", description: "Codex discovery before bootstrap", check: invCodex003 },
];

export function exportStaticInvariants() {
  return PRESET_INVARIANTS.map((inv) => ({
    invariant_id: inv.invariant_id,
    category: inv.category,
    description: inv.description,
  }));
}
